src/function/scalar/date/date_trunc.cpp

```diff
@@ -692,6 +692,9 @@ static unique_ptr<FunctionData> DateTruncBind(ClientContext &context, ScalarFunc
 	case DatePartSpecifier::DOY:
 		switch (arguments[1]->return_type.id()) {
```

Comment:
I think the problem here is actually that you are looking at the argument types to begin with - date trunc does not accept `timestamp` parameters (besides `timestamp`), so a cast will be automatically added. The cast will only be performed **after** the bind, however. This function should either be templated, or look at the `arguments` inside the `bound_function`.

Reply:
Ah so you are saying that instead of "knowing" all the types that could be cast to `TIMESTAMP` it should discover what that argument will be by looking at the original binding types?

Reply:
Exactly